File: src/main/java/com/ragin/bdd/cucumber/hooks/ResetHooks.java
package com.ragin.bdd.cucumber.hooks;

import com.ragin.bdd.cucumber.core.DatabaseExecutorService;
import com.ragin.bdd.cucumber.core.ScenarioStateContext;
import io.cucumber.java.Before;
import io.cucumber.java.Scenario;
import lombok.extern.apachecommons.CommonsLog;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;

@CommonsLog
public class ResetHooks {
    private static final String RESET_DATABASE_FILE = "database/reset_database.xml";

    @Autowired
    DatabaseExecutorService databaseExecutorService;

    /**
     * Execute the {@link #RESET_DATABASE_FILE} liquibase script to reset the database.
     *
     * @param scenario      Current Cucumber Scenario
     * @throws Exception    Not able to execute the database script
     */
    @Before(order = 2)
    @Transactional
    public void resetDatabase(final Scenario scenario) throws Exception {
        // Check if the file exist
        if (getClass().getClassLoader().getResource(RESET_DATABASE_FILE) == null) {
            log.debug(String.format("No reset database file (%s) found.", RESET_DATABASE_FILE));
        } else {
            // Execute the liquibase script to reset the DB
            log.info(String.format("Reset database for scenario %s", scenario.getName()));
            databaseExecutorService.executeLiquibaseScript(RESET_DATABASE_FILE);
        }
    }

    /**
     * Reset state of BaseCucumberCore
     * @param scenario  Current Cucumber Scenario
     */
    @Before(order = 3)
    public void resetBaseCucumberCoreState(final Scenario scenario) {
        log.info(String.format("Cleanup test state for scenario %s", scenario.getName()));
        ScenarioStateContext.current().reset();
    }
}

File: src/main/kotlin/com/ragin/bdd/cucumber/hooks/ResetHooks.kt
package com.ragin.bdd.cucumber.hooks

import com.ragin.bdd.cucumber.core.IDatabaseExecutorService
import com.ragin.bdd.cucumber.core.ScenarioStateContext.addJsonIgnoringArrayOrder
import com.ragin.bdd.cucumber.core.ScenarioStateContext.addJsonIgnoringExtraArrayElements
import com.ragin.bdd.cucumber.core.ScenarioStateContext.addJsonIgnoringExtraFields
import com.ragin.bdd.cucumber.core.ScenarioStateContext.reset
import io.cucumber.java.Before
import io.cucumber.java.Scenario
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.transaction.annotation.Transactional

open class ResetHooks(private val databaseExecutorService: IDatabaseExecutorService) {
    companion object {
        private const val RESET_DATABASE_FILE = "database/reset_database.xml"
        private val log = KotlinLogging.logger { }
    }

    /**
     * Execute the [.RESET_DATABASE_FILE] liquibase script to reset the database.
     *
     * @param scenario      Current Cucumber Scenario
     * @throws Exception    Not able to execute the database script
     */
    @Before(order = 2)
    @Transactional
    @Throws(Exception::class)
    open fun resetDatabase(scenario: Scenario) {
        // Check if the file exist
        if (javaClass.classLoader.getResource(RESET_DATABASE_FILE) == null) {
            log.debug { "No reset database file ($RESET_DATABASE_FILE) found." }
        } else {
            // Execute the liquibase script to reset the DB
            log.info { "Reset database for scenario ${scenario.name}" }
            databaseExecutorService.executeLiquibaseScript(RESET_DATABASE_FILE)
        }
    }

    /**
     * Reset state of BaseCucumberCore
     * @param scenario  Current Cucumber Scenario
     */
    @Before(order = 3)
    fun resetBaseCucumberCoreState(scenario: Scenario) {
        log.info { "Cleanup test state for scenario ${scenario.name}" }
        reset()
    }

    @Before(order = 10, value = "@bdd_lib_json_ignore_new_array_elements")
    fun jsonIgnoreExtraArrayElements(scenario: Scenario) {
        log.info { "Set JSON ignore for extra array elements for scenario ${scenario.name}" }
        addJsonIgnoringExtraArrayElements()
    }

    @Before(order = 10, value = "@bdd_lib_json_ignore_extra_fields")
    fun jsonIgnoreNewFields(scenario: Scenario) {
        log.info { "Set JSON ignore for extra fields for scenario ${scenario.name}" }
        addJsonIgnoringExtraFields()
    }

    @Before(order = 10, value = "@bdd_lib_json_ignore_array_order")
    fun jsonIgnoreArrayOrder(scenario: Scenario) {
        log.info { "Set JSON ignore array order for scenario ${scenario.name}" }
        addJsonIgnoringArrayOrder()
    }
}
